### PR TITLE
Added Gradient Centralization example

### LINKED ISSUE #418: Regarding adding an example on Gradient Centralization
I plan to propose an example demonstrating an implementation of the paper [Gradient Centralization: A New Optimization Technique for Deep Neural Networks](https://arxiv.org/abs/2004.01461) which can both speed up the training process and improve the final generalization performance of DNNs especially for general image classification, fine-grained image classification, detection and segmentation and Person ReID.

I have also brought this to TensorFlow as a [ready to use Python package](https://github.com/Rishit-dagli/Gradient-Centralization-TensorFlow) to use with Keras. Here is an example notebook from this repository [Open In Colab](https://colab.research.google.com/github/Rishit-dagli/Gradient-Centralization-TensorFlow/blob/main/examples/gctf_horses_v_humans.ipynb) and I plan to propose a similar one with any changes if required.

### PR DESCRIPTION
**Pull request by Rishit-dagli**
This example implements [Gradient Centralization](https://arxiv.org/abs/2004.01461), a new optimization technique for Deep Neural Networks by Yong et al., and demonstrates it on Laurence Moroney's [Horses or Humans Dataset](https://www.tensorflow.org/datasets/catalog/horses_or_humans). Gradient Centralization can both speed up the training process and improve the final generalization performance of DNNs especially for general image classification, fine-grained image classification, detection and segmentation, and Person ReID. It operates directly on gradients by centralizing the gradient vectors to have zero mean.

The below table shows the improvement this example demonstrates on using Gradient Centralization on the mentioned dataset, in 10 epochs:

|              | Loss   | Accuracy | Training Time |
|--------------|--------|----------|---------------|
| Not using GC | 0.5489 | 75.94    | 185.98s       |
| Using GC     | 0.4069 | 82.47    | 164.58s       |

This same example could also be tried out in this [Colab Notebook](https://colab.research.google.com/drive/1S68JYKnhJs69s1sptkYqVqk8N8SnP8lx?usp=sharing).

cc: @sayakpaul 

---

Closes #418

### CODE REVIEW

File: examples/vision/gradient_centralization.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+to use Gradient Centralization.
+
+The experiments in [this paper](https://arxiv.org/abs/2004.01461) on various
+applications, including general image classification, fine-grained image classification,
+detection and segmentation and Person ReID demonstrate that GC can consistently improve
+the performance of DNN learning.
+"""
+
+optimizer = RMSprop(learning_rate=1e-4)
+optimizer.get_gradients = gctf.centralized_gradients_for_optimizer(optimizer)
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:206]: I see the implementation is pretty short: https://github.com/Rishit-dagli/Gradient-Centralization-TensorFlow/blob/a7c5226dad86ca42341061e3fafc8c8d1ec3f51f/gctf/centralized_gradients.py#L5-L42

I recommend packaging the implementation directly in the example. The goal of a code example is to explain how to implement something, not just demonstrate how to use an external package.

The easiest way to implement this would be to create a subclass of `RMSprop` that overrides the `get_gradients` method. You can make it very short by leveraging `super().get_gradients()`.

File: examples/vision/gradient_centralization.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        layers.Conv2D(64, (3, 3), activation="relu"),
+        layers.MaxPooling2D(2, 2),
+        layers.Conv2D(64, (3, 3), activation="relu"),
+        layers.MaxPooling2D(2, 2),
+        layers.Flatten(),
+        layers.Dropout(0.5),
+        layers.Dense(512, activation="relu"),
+        layers.Dense(1, activation="sigmoid"),
+    ]
+)
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:148]: Please print a summary. Also please let me know what the summary looks like (do you have a live notebook?)

File: examples/vision/gradient_centralization.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+print(f"Training images: {metadata.splits['train'].num_examples}")
+print(f"Test images: {metadata.splits['test'].num_examples}")
+
+"""
+# Use Data Augmentation
+
+We will rescale the data to `[0, 1]`  andperform simple augmentations to our data.
+"""
+
+rescale = tf.keras.Sequential([layers.experimental.preprocessing.Rescaling(1.0 / 255)])
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:81]: You don't need a Sequential model here. Just `rescale = layers.experimental.preprocessing.Rescaling(1.0 / 255)`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! Looks very interesting. Main comment here is that you should put the GC implementation in the code example directly. Feel free to link to the other package in the text paragraphs, but don't depend on it.

File: examples/vision/gradient_centralization.py (hunk 1 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py:206]: Sure thing, that makes a lot of sense, thanks!
I added that but did not use `super().get_gradients()` since I did not want to run the `get_gradients()` function first.

File: examples/vision/gradient_centralization.py (hunk 2 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py:148]: Added that. By "a live notebook" do you mean something like a runnable Colab? If yes here it is [Colab Notebook](https://colab.research.google.com/drive/1S68JYKnhJs69s1sptkYqVqk8N8SnP8lx?usp=sharing).

File: examples/vision/gradient_centralization.py (hunk 3 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py:81]: Ah yes, thanks!

**Comment by Rishit-dagli** [comment]
Thanks so much, @fchollet for reviewing the example and super helpful reviews, I have implemented them.

File: examples/vision/gradient_centralization.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+# Setup
+"""
+
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow.keras import layers
+from tensorflow.keras.layers import Conv2D
+from tensorflow.keras.layers import MaxPooling2D
+from tensorflow.keras.layers import Dropout
+from tensorflow.keras.layers import Dense
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:44]: Prefer just importing `layers` then accessing module members

File: examples/vision/gradient_centralization.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+from tensorflow.keras.optimizers import RMSprop
+import keras.backend as K
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:46]: You don't actually need this module, just us `tf.something` when you need a math op.

File: examples/vision/gradient_centralization.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+data_augmentation = tf.keras.Sequential(
+    [
+        layers.experimental.preprocessing.RandomFlip("horizontal_and_vertical"),
+        layers.experimental.preprocessing.RandomRotation(0.3),
+        layers.experimental.preprocessing.RandomZoom(0.2),
+    ]
+)
+
+batch_size = 128
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:89]: You can move these 2 params to be together with the other ones above

File: examples/vision/gradient_centralization.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+column vectors of the weight matrix, and then remove the mean from each column vector.
+
+The experiments in [this paper](https://arxiv.org/abs/2004.01461) on various
+applications, including general image classification, fine-grained image classification,
+detection and segmentation and Person ReID demonstrate that GC can consistently improve
+the performance of DNN learning.
+"""
+
+
+class gc_rmsprop(RMSprop):
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:210]: To improving the pacing of the tutorial, move the optimizer definition (and explanations of what GC is) to be right after model definition (callbacks should be defined afterwards, before training)

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by fchollet** [comment]
Also, please don't include any files other than the `.py` file in the PR until the PR is approved -- they'll have to be regenerated anyway.

File: examples/vision/gradient_centralization.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    def get_gradients(self, loss, params):
+        # We here just provide a modified get_gradients() function since we are
+        # trying to just compute the centralized gradients.
+
+        grads = []
+        for grad in K.gradients(loss, params):
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Sure thing, also since we are making a subclass of `tf.keras.optimizer.RMSProp` and only care about the `get_gradients()` method as you mentioned earlier I have used `super().get_gradients()` to make it even more concise.

File: examples/vision/gradient_centralization.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+            grad_len = len(grad.shape)
+            if grad_len > 1:
+                axis = list(range(grad_len - 1))
+                grad -= tf.reduce_mean(grad, axis=axis, keep_dims=True)
+            grads.append(grad)
+
+        if None in grads:
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: As I was talking in [this comment](https://github.com/keras-team/keras-io/pull/497#discussion_r652634579), I use `super().get_gradients()` so though we don't need it, I think this check would already be done for us through `super().get_gradients()`

File: examples/vision/gradient_centralization.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Ah sorry, renamed the class.

File: examples/vision/gradient_centralization.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+            raise ValueError(
+                "An operation has `None` for gradient. "
+                "Please make sure that all of your ops have a "
+                "gradient defined (i.e. are differentiable). "
+                "Common ops without gradient: "
+                "K.argmax, K.round, K.eval."
+            )
+        if hasattr(optimizer, "clipnorm") and optimizer.clipnorm > 0:
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Got it, added some more text to support what you mention, and removed gradient clipping functionality.

File: examples/vision/gradient_centralization.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+be installed with this command:
+
+```
+pip install tensorflow-datasets
+```
+
+We will be implementing Gradient Centralization in this example but you could also use
+this very easily with a package I built,
+[gradient-centralization-tf](https://github.com/Rishit-dagli/Gradient-Centralization-TensorFlow).
+[gradient-centralization-tf](https://github.com/Rishit-dagli/Gradient-Centralization-TensorFlow).
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Sorry, this was a typo from my end.

File: examples/vision/gradient_centralization.py (hunk 5 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Yep, after removing gradient clipping we only have 1 math op for which we now use `tf.reduce_mean()`

File: examples/vision/gradient_centralization.py (hunk 6 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Indeed having the params together would make a lot more sense.

File: examples/vision/gradient_centralization.py (hunk 7 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: I have made these edits, however, may I request you to also review the additional text I added to support this:

https://github.com/keras-team/keras-io/blob/52c5632c63215a59244f5dc2b0ebe7dc7c421d94/examples/vision/gradient_centralization.py#L158-L161

and also:

https://github.com/keras-team/keras-io/blob/52c5632c63215a59244f5dc2b0ebe7dc7c421d94/examples/vision/gradient_centralization.py#L185-L189

**Comment by Rishit-dagli** [comment]
@fchollet Thanks for all the reviews, I have edited the tutorial accordingly.

> Also, please don't include any files other than the `.py` file in the PR until the PR is approved -- they'll have to be regenerated anyway.

About this, earlier I had added the notebook at `examples/vision/ipynb/gradient_centralization.ipynb` and now have removed that.

File: examples/vision/gradient_centralization.py [hunk 12]
```diff
@@ -0,0 +1,264 @@
+"""
+Title: Gradient Centralization for Better Training Performance
+Author: [Rishit Dagli](https://github.com/Rishit-dagli)
+Date created: FILLME
+Last modified: FILLME
```
Comment (fchollet) [inline examples/vision/gradient_centralization.py:5]: Update these fields

**Review by fchollet — COMMENTED** [review]
Thanks for the update. Looks great! Please add the autogenerated files.

File: examples/vision/gradient_centralization.py (hunk 12 above)
Comment (Rishit-dagli) [reply examples/vision/gradient_centralization.py]: Got it.

**Comment by Rishit-dagli** [comment]
@fchollet I have added the autogenerated files.

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution! 👍
